packages/ckeditor5-watchdog/src/editorwatchdog.ts

```diff
@@ -284,13 +295,31 @@ export default class EditorWatchdog<TEditor extends Editor = Editor> extends Wat
 	}
 
 	/**
-	 * Returns the editor data.
+	 * Gets the all data that are required to reinitialize editor instance.
 	 */
-	private _getData(): Record<string, string> {
-		const data: Record<string, string> = {};
+	private _getData(): EditorData {
+		const editor = this.editor!;
+		const rootNames = editor.model.document.getRootNames();
+		const data: EditorData = {
+			roots: {},
+			markers: {}
+		};
+
+		rootNames.forEach( rootName => {
+			const root = editor.model.document.getRoot( rootName )!;
+
+			data.roots[ rootName ] = {
+				content: JSON.stringify( Array.from( root.getChildren() ) ),
+				attributes: Array.from( root.getAttributes() )
```

Comment:
One thing I am worried about is that if there is an object saved as an attribute, you will get a reference to this object. I am not sure if `getAttributes()` or setting attributes through writer handles it in any way. I am afraid that a reference to this object may cause:

1.  Unexpected changes if the original object got changed somewhere.
2.  Memory leaks, if the original object had a reference to something.

In general, people should not use objects as attributes and surely not objects that keep non-primitve data types (other than maybe more plain objects, you get the point). But maybe to be perfectly safe, we should also stringify/parse this data?